PYME/cluster/rules.py

```diff
@@ -0,0 +1,616 @@
+        self.start_at = startAt
+        self.serverfilter = serverfilter
+    
+        #load data source
+        if dataSourceModule is None:
+            DataSource = DataSources.getDataSourceForFilename(dataSourceID)
+        else:
+            DataSource = __import__('PYME.IO.DataSources.' + dataSourceModule,
+                                    fromlist=['PYME', 'io', 'DataSources']).DataSource #import our data source
+    
+        self.ds = DataSource(self.dataSourceID)
+    
+        logger.debug('DataSource.__class__: %s' % self.ds.__class__)
+    
+    def _task_template(self, context):
+        tt = {'id': '{{ruleID}}~{{taskID}}',
+              'type': 'localization',
+              'taskdef': {'frameIndex': '{{taskID}}', 'metadata': self.results_md_uri},
+              'inputs': {'frames': self.dataSourceID},
+              'outputs': {'fitResults': self.worker_resultsURI + '/FitResults',
+                          'driftResults': self.worker_resultsURI + '/DriftResults'}
+              }
+        return json.dumps(tt)
+
+    def prepare(self):
+        """
+        Do any setup work - e.g. uploading metadata required before the rule is triggered
+
+        Returns
+        -------
+
+        post_args : dict
+            a dictionary with arguments to pass to RulePusher._post_rule() - specifically timeout, max_tasks, release_start, release_end
+
+        """
+        #set up results file:
+        logging.debug('resultsURI: ' + self.worker_resultsURI)
+        clusterResults.fileResults(self.worker_resultsURI + '/MetaData', self.mdh)
+        clusterResults.fileResults(self.worker_resultsURI + '/Events', self.ds.getEvents())
+
+        # set up metadata file which is used for deciding how to launch the analysis
+        clusterIO.put_file(self.resultsMDFilename, self.mdh.to_JSON().encode(), serverfilter=self.serverfilter)
+
+        #wait until clusterIO caches clear to avoid replicating the results file.
+        #time.sleep(1.5) #moved inside polling thread so launches will run quicker
+
+        self.currentFrameNum = self.start_at
+        numTotalFrames = self.ds.getNumSlices()
+        self.frames_outstanding=numTotalFrames - 1 - self.currentFrameNum
+        
+        return {}
+
+    @property
+    def complete(self):
+        """
+        Is this rule complete, or do we need to poll for more input?
+
+        Over-ridden in localisation rule
+        Returns
+        -------
+
+        """
+        return self.ds.is_complete and not (self.frames_outstanding  > 0)
+
+    def get_new_tasks(self):
+        """
+        Over-ridden in rules where all the data is not gauranteed to be present when the rule is created
+
+        Returns
+        -------
+
+        release_start, release_end : the indices of starting and ending tasks to release
+
+        """
+        numTotalFrames = self.ds.getNumSlices()
+        logging.debug('numTotalFrames: %s, currentFrameNum: %d' % (numTotalFrames, self.currentFrameNum))
```

Comment:
@David-Baddeley have a look at _rules for a 0-indexed version of all this which ought to fix taskCompleted == (tasksPosted - 1)

startAt is zero indexed, as is the task array we use on the ruleserver, all of these +1 and -1 can (and should) go

Reply:
I'm not sure I follow - most of these +1s are needed  and correct for **0-based** indexing. e.g. 

- have a total of 15 frames, all app pushed => `numTotalFrames == 15`, `currentFrameNum ==14` ( ie `nFrames = frameNum +1`) which makes line 542 correct
- ditto for line 454 - 15 frames makes the frameNum for the last frame 14.
- and for L 548 - if your current frame number is 14 then you have released 15 frames.
- finally for the return - these are the values to pass to the release call, which uses slice notation - i.e. the last frame of the range is excluded

Reply:
I see you got it sorted in the latest commit. Thanks David. Anything holding this up before merging?